packages/extension-ui/src/hooks/useRequestsPagination.ts

```diff
@@ -0,0 +1,28 @@
+// Copyright 2019-2023 @polkadot/extension-ui authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import { useEffect, useState } from 'react';
+
+const getSafeIndex = (index: number, arrLength: number) => Math.min(index, Math.max(arrLength - 1, 0));
+
+const useRequestsPagination = <T>(requests: Array<T>) => {
+  const [index, setIndex] = useState(0);
+
+  const next = () => setIndex((requestIndex) => Math.min(requestIndex + 1, requests.length));
+  const previous = () => setIndex((requestIndex) => Math.max(requestIndex - 1, 0));
+
+  useEffect(() => {
+    setIndex((currIndex) => getSafeIndex(currIndex, requests.length));
+  }, [requests]);
+
+  const finalIndex = getSafeIndex(index, requests.length);
```

Comment:
I don't think you need this - seems to me that `index`'s safety is guaranteed by the above calculations, isn't it?

If no, ten I'd pick a different name thatn `finalIndex` - I was thinking that it's as in _last index_, until I read the code twice. Maybe `safeIndex` or `validatedIndex` or `sanitizedIndex`.

Reply:
I think there might be a short window where `index` is out of bound:
* requests.length change from 3 -> 2
* this triggers useEffect setting new `index` value after component is re-rendered
* index is updated, but one re-render later, which shouldn't hurt anyone, but I think it's safer this way